### DISCUSSION TITLE
`locate` feels very unintuitive (Or am I just using the wrong function?)

### DISCUSSION THREAD

**Discussion by ParadaCarleton** [typst/typst#1361] [Q&A] [answered]
### Description

I assume there's something I'm not getting here, but I don't know what. From the name and documentation, it feels like the point of the `locate` function is to find the location or position of an element. If that's the case, I'm confused as to why it:
1. Takes a function as an input (rather than a block, or just not taking arguments at all),
2. Returns a content block, instead of just a pair of coordinates, and
3. Apparently it's impossible to convert this content into a pair of integers?

Am I misunderstanding what this function is supposed to do? Is there another function that just returns the current position as a pair of lengths?

### Use Case

\

**Comment by maxcrees** [comment, marked as answer]
So `locate` has a somewhat unusual design, but once you understand it and use it in practice it starts to make sense. 

The idea is that if it were to just return the information directly, then that information could be invalidated as the rest of your document is compiled (things move around etc.). When that happens, the compiler needs to recalculate everything that depends on that information.

By making the function expect a function as a parameter, it nicely encapsulates the piece of your document that depends on that information.

The function that you write and give to `locate` in turn expects a single parameter of its own. You can call this whatever you like, but the docs like to use the name `loc`. It is this object that you can inspect for coordinates, current page number, and the like.

`locate` itself instead returns the content from whatever you put inside of the function you write within it. This could be nothing (no content), but even then that nothing is still placed in the document somewhere so that the compiler knows what location information to provide to your function.

**Comment by maxcrees** [comment]
In that sense `locate` is more about finding the **current** position in the document. If you want to find the position of a particular part of the document, you use it in tandem with `query`.
